### ISSUE TITLE
[Feature] Multi-GPU support for PACMAP

### ISSUE THREAD

**Issue by huguesva** [TorchDR/TorchDR#242]
## Summary

PACMAP currently does not support multi-GPU execution. This issue documents the technical challenges and potential solutions.

## Background

PACMAP uses three types of point pairs in its loss function:
- **NB (Near)**: k-nearest neighbors in input space
- **MN (Mid-Near)**: dynamically sampled medium-distance pairs
- **FP (Far Pairs)**: randomly sampled negative pairs

While NB and FP follow patterns similar to UMAP (which supports multi-GPU), the MN component presents unique challenges.

### How Mid-Near Sampling Works

At **every training iteration**, PACMAP:
1. For each point, randomly samples 6 candidate indices from the **entire dataset**
2. Computes distances to candidates in **input space** (requires access to `X`)
3. Selects the 2nd closest as the mid-near pair

```python
# Simplified from pacmap.py
for i in range(n_mid_near):
    candidates = torch.randint(0, n_samples, (n_samples, 6))  # global random indices
    D = pairwise_distances_indexed(self.X_, key_indices=candidates)  # needs full X
    mid_near_indices[:, i] = kmin(D, k=2)[1][:, 1]  # pick 2nd closest
```

### Why This Breaks Multi-GPU

In distributed mode with data sharded across GPUs:

| GPU 0 | GPU 1 | GPU 2 | GPU 3 |
|-------|-------|-------|-------|
| Points 0-999 | Points 1000-1999 | Points 2000-2999 | Points 3000-3999 |

When Point 0 (GPU 0) samples candidates `[42, 1337, 2500, 3800]`:
- Index 42 → GPU 0 (local)
- Index 1337 → GPU 1 (remote)
- Index 2500 → GPU 2 (remote)
- Index 3800 → GPU 3 (remote)

Computing distances requires fetching `X[1337]`, `X[2500]`, `X[3800]` from other GPUs **at every iteration**.

### Comparison with UMAP/TSNE

| Component | Data Access | When | Multi-GPU |
|-----------|-------------|------|-----------|
| UMAP neighbors | Sparse affinity | Once (precomputed) | Supported |
| UMAP negatives | Embedding only | Every step | Supported |
| PACMAP NB | Sparse affinity | Once (precomputed) | Feasible |
| PACMAP FP | Embedding only | Every step | Feasible |
| **PACMAP MN** | **Full input X** | **Every step** | **Challenging** |

### Option 1: Replicate X on all GPUs
- **Pros**: Simple, no algorithm changes
- **Cons**: Memory inefficient, limits scalability

### Option 2: All-gather for mid-near candidates
- **Pros**: Correct semantics
- **Cons**: Communication overhead every iteration

### Option 3: Local-only mid-near sampling
- **Pros**: No communication, fast
- **Cons**: Changes algorithm behavior (samples only from local shard)

### Option 4: Precompute mid-near pairs
- **Pros**: One-time communication cost
- **Cons**: Not faithful to original PACMAP (which uses dynamic sampling)

## References

- PACMAP paper: https://arxiv.org/abs/2012.04456
- Current blocking code: `torchdr/neighbor_embedding/pacmap.py:120-121`
- PACMAPAffinity blocking code: `torchdr/affinity/knn_normalized.py:549-552`

**Comment by huguesva** [comment]
The *current* unsupported behavior (PACMAP/PACMAPAffinity raise on a truthy `distributed`) is now documented and locked by regression tests in #329 (issue #328). This issue remains open for the actual multi-GPU implementation, whose main challenge is the per-iteration mid-near sampling over the full input as described above.
